<div style="border:solid purple 2px; padding: 20px"> 

Мария, привет! 👋 Поздравляю с выполненным проектом 🙂

Меня зовут Рустам Муртазин, и я буду делать ревью проектной работы по предобработке данных. Давай будем общаться на **«ты»**. Если это неприемлемо, обязательно напиши мне в комментариях — мы перейдем на **«вы»**.

Ты проделала большую работу над проектом, однако в любом деле всегда можно можно что-то улучшить. Моя основная цель — не показать, что где-то совершены какие-либо ошибки, а поделиться своим опытом, который поможет тебе в дальнейших работах. Далее в файле ты сможешь увидеть мои комментарии, постарайся учесть их при выполнении следующих заданий.

Обрати внимание в первую очередь на те, что помечаны <span style="color:red">красным цветом</span>. После их доработки проект будет принят. <span style="color:green">Зеленым цветом</span> отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. <span style="color:orange">Оранжевым цветом</span> выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно). Также в проекте могут быть небольшие «лайфхаки» по Python, не относящиеся к проекту, их я выделил в фиолетовую рамку)

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее».

В целом работа хорошая, видно, что ты старалась. Дорабатывай комментарии и отправляй работу на повторное ревью, я буду ждать 😉

<span style="color:magenta">Привет, Рустам! Спасибо за твое ревью, мне понравилось :). Я полностью согласна с твоими замечаниями по оптимизации решений, честно говоря, где-то мой "внутренний программист" ликует, что схалтурить не получилось и надо сделать лучше. Я не стала удалять предыдущий вариант решения, возможно так будет удобней для проверки, буду признательна, если ты укажешь, как лучше поступать в следующий раз, удалять или нет. Решения по оптимизации я поместила сразу после предыдущего варианта решения и перед выводом. Все мои новые комментарии выделены этим жутким цветом :)</span>

<div class="alert alert-info">
<h2> Комментарий ревьюера v2 <a class="tocSkip"> </h2>

Решение оставить предыдущий вариант очень правильное. Дальше тоже оставляй, как и реализации, в которых возникли ошибки и что-то не получилось, так легче понять в чем проблема)
    
Кстати, розовый цвет мне нравится, а можно вот так оформлять еще)

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Откройте файл с данными и изучите общую информацию

<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Ты уже наверное заметила, недавно добавили плагин `toc` для Jupyter, позволяющий сделать [автоматическое оглавление](https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/toc2/README.html). Так вот, если тебе не нравится автоматическая нумерация с `toc`, то ее можно убрать. Для этого нужно сделать следующее:

- выведи оглавление, для этого нажми на значок списка (находится справа от `Validate`)
- слева выйдет само оглаление, нажми на шестиренку
- и убери галку с функции автоматической нумерации `Automatically number headings`


In [1]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

**Вывод:**

Мы видим большой массив данных, который, очевидно, нуждается в предобработке для дальнейшего проведения исследования.

<div style="border:solid purple 5px; padding: 20px"> 
<h2 align="center"> Рубрика «Питонячий лайфхакер» <a class="tocSkip"> </h2>

<h3> PEP 8 - руководство по написанию кода на Python  <a class="tocSkip"> </h3>

В python есть своя философия. Именно отсюда берутся определенные правила и стандарты. В пайтон такое соглашение назвали PEP 8. PEP 8 создан на основе рекомендаций Гуидо ван Россума (создатель языка Python) с добавлениями от Барри Уорсо (соратник).

Единый стиль оформления делает код понятным для самого программиста и его коллег с разным уровнем подготовки. В идеале наиболее сложный фрагмент кода должен быть понятен с первого прочтения. Это упрощает командную разработку и обучение новичков, позволяет вам быстро возвращаться к собственным давним проектам.

Каждый уважающий себя питонист обязан знать этот стандарт. Читайте [оригинал](https://www.python.org/dev/peps/pep-0008/) или [перевод](https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html), главное  —  вдумчиво!

Когда проект небольшой, несложно следить за качеством кода, но вот когда проект принимает маштабы ... часто это становиться проблемой. В этом случае выручат скрипты автоматической проверки кода. Кстати, PyCharm проверяет написанное «на лету». А если вы привыкли работать в Jupyter, то на GitHub есть целый раздел [Python Code Quality Authority](https://github.com/PyCQA/), где хранятся утилиты для повышения качества кода, в том числе для проверки стиля на соответствие PEP 8: flake8, pycodestyle, pep8-naming.

![](https://i.ibb.co/Gv0PvHQ/mini-magick20171220-84-vtyy26.png)

## Шаг 2. Предобработка данных

### Обработка пропусков

Посмотрим информацию о данных нашей таблицы:

In [2]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Обнаружены пропущенные значения в столбцах *'days_employed'* и *'total_income'*. Проверка выявила,что все пропуски представлены как NaN:

In [3]:
data['days_employed'].isnull().sum()

2174

In [4]:
data['total_income'].isnull().sum()

2174

Проверим, связаны ли эти пропуски между собой:

In [5]:
data.loc[(data['days_employed'].isnull() == True)&(data['total_income'].isnull() == True)]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Пропуски синхронны по двум столбцам, что позволяет сделать вывод, что они не случайны. Проверим, есть ли связь между количеством пропусков в этих столбцах с данными столбца *'income_type'*. Для начала посмотрим, какие данные он содержит, создадим список категорий:

In [6]:
income_type_categories = data['income_type'].unique()
income_type_categories

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

Создадим функцию для подсчета процентного содержания записей с пропусками для каждой категории из столбца *'income_type'*:

In [7]:
def procent_part(data, income_type):
    nan_count = data.loc[(data['days_employed'].isnull() == True)&(data['income_type'] == income_type), 'income_type'].count()
    all_count = data.loc[data['income_type'] == income_type, 'income_type'].count()
    procent = nan_count/all_count
    return 'Пропусков в категории "' + income_type + '" {:.1%}'.format(procent)    

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Отличная и полезная функция 👏

Приведем результаты работы функции для каждой категории из списка *income_type_categories*: 

In [8]:
for income_type in income_type_categories:
    print(procent_part(data, income_type))

Пропусков в категории "сотрудник" 9.9%
Пропусков в категории "пенсионер" 10.7%
Пропусков в категории "компаньон" 10.0%
Пропусков в категории "госслужащий" 10.1%
Пропусков в категории "безработный" 0.0%
Пропусков в категории "предприниматель" 50.0%
Пропусков в категории "студент" 0.0%
Пропусков в категории "в декрете" 0.0%


Таким образом, мы выяснили, что почти в каждой категории клиентов есть записи с пропусками в столбцах о стаже и доходе. Информация о стаже(столбец 'days_employed') для нашего исследования не используется, поэтому для простоты можно заполнить пропуски в столбце 'days_employed' нулями: 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Верно, или можно было оставить как есть

In [9]:
data['days_employed'] = data['days_employed'].fillna(0)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Тип данных float в столбце *'days_employed'* удобно заменить на целочисленный. Так как встречаются отрицательные значения в этом столбце, необходимо исправить это с помощью функции abs():

In [10]:
data['days_employed'] = abs(data['days_employed'].astype('int'))

Проверим, содержит ли столбец *'total_income'* отрицательные значения. Найдем минимальное и максимальное значение из этого столбца:

In [11]:
data['total_income'].min()

20667.26379327158

In [12]:
data['total_income'].max()

2265604.028722744

Как видим, отрицательных значений в столбце нет.

В связи с тем, то объем данных с пропущенными значениями по столбцу *'total_income'* довольно большой и встречается почти у всех категорий клиентов по типу занятости (столбец *'income_type'*), мы не можем удалять или замещать нулями эти пропуски, так как это может существенно отразится на дальнейших результатах исследования (вопрос о зависимости кредитной надежности клиента от его дохода). Поэтому правильнее будет заменить пропущенные значения столбца *'total_income'* на его медианное значение по каждой из категорий клиентов из списка *income_type_categories*. 
Создадим метод, который будет считать медиану для каждой категории клиентов и заполнять пропущенные значения столбца *'total_income'* для этих клиентов. Метод также заполнит список из значений медианы для каждой категории клиентов:

In [13]:
def income_type_mediana(data, income_type, median_list):
    data_income_type = data.loc[data['income_type'] == income_type]
    median = data_income_type['total_income'].median()
    data.loc[data['income_type'] == income_type, 'total_income'] = data.loc[data['income_type'] == income_type, 'total_income'].fillna(median)
    median_list.append(median)

Применим его к каждому элементу списка *income_type_categories* и посмотрим, какие медианы у нас получились для каждой из категории списка: 

In [14]:
median_list = []
for income_type in income_type_categories:
    income_type_mediana(data, income_type, median_list)
print(median_list)    

[142594.39684740017, 118514.48641164352, 172357.95096577113, 150447.9352830068, 131339.7516762103, 499163.1449470857, 98201.62531401133, 53829.13072905995]


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Хорошая реализация! Вот еще один вариант, как можно было сделать замену, но уже не используя циклы и отдельные функции 😉
    
    data['total_income'] = data.groupby('income_type')['total_income'].apply(lambda x: x.fillna(x.median()))

</div>

Теперь удобно заменить вещественный тип данных в столбце *'total_income'* на целочисленный:

In [15]:
data['total_income'] = data['total_income'].astype('int')

Проверим результаты работы нашего метода:

In [16]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


Проверим также остальные столбцы на предмет их содержания:

In [17]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

Заменим в столбце *'children'* отрицательные значения на абсолютные:

In [18]:
data['children'] = abs(data['children'])
data['children'].unique()

array([ 1,  0,  3,  2,  4, 20,  5])

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip">  </h2>

Мне нравится ход твоих мыслей и детальная проработка 👍

</div>

Проверяем следующие столбцы:

In [19]:
data['dob_years'].unique()

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51,  0, 59, 29, 60, 55, 58, 71, 22, 73,
       66, 69, 19, 72, 70, 74, 75])

In [20]:
data['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Здесь для удобства следует привести все строки к нижнему регистру:

In [21]:
data['education'] = data['education'].str.lower()
data['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Проверяем следующие столбцы:

In [22]:
data['education_id'].unique()

array([0, 1, 2, 3, 4])

In [23]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

Приведем строки к нижнему регистру:

In [24]:
data['family_status'] = data['family_status'].str.lower()
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'не женат / не замужем'], dtype=object)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip">  </h2>

Молодец, что приводишь данные к общему виду, правильное решение

Смотрим дальше:

In [25]:
data['family_status_id'].unique()

array([0, 1, 2, 3, 4])

In [26]:
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [27]:
data['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

In [28]:
data['debt'].unique()

array([0, 1])

In [29]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

С остальными столбцами, кроме столбца *'purpose'*, кажется, все в порядке. С ним разберемся ниже.

### Обработка дубликатов

Посчитаем количество дубликатов:

In [30]:
data.duplicated().sum() 

71

Посмотрим, какие строки дублируются:

In [31]:
data.loc[data.duplicated() == True]

children  days_employed  dob_years education  education_id  \
2849          0              0         41   среднее             1   
3290          0              0         58   среднее             1   
4182          1              0         34    высшее             0   
4851          0              0         60   среднее             1   
5557          0              0         58   среднее             1   
...         ...            ...        ...       ...           ...   
20702         0              0         64   среднее             1   
21032         0              0         60   среднее             1   
21132         0              0         47   среднее             1   
21281         1              0         30    высшее             0   
21415         0              0         54   среднее             1   

          family_status  family_status_id gender income_type  debt  \
2849    женат / замужем                 0      F   сотрудник     0   
3290   гражданский брак                 1      F   пенсионер     0   
4182   гражданский брак                 1      F   сотрудник     0   
4851   гражданский брак                 1      F   пенсионер     0   
5557   гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
20702   женат / замужем                 0      F   пенсионер     0   
21032   женат / замужем                 0      F   пенсионер     0   
21132   женат / замужем                 0      F   сотрудник     0   
21281   женат / замужем                 0      F   сотрудник     0   
21415   женат / замужем                 0      F   пенсионер     0   

       total_income                            purpose  
2849         142594            покупка жилья для семьи  
3290         118514                    сыграть свадьбу  
4182         142594                            свадьба  
4851         118514                            свадьба  
5557         118514                    сыграть свадьбу  
...             ...                                ...  
20702        118514         дополнительное образование  
21032        118514              заняться образованием  
21132        142594                       ремонт жилью  
21281        142594  покупка коммерческой недвижимости  
21415        118514                  операции с жильем  

[71 rows x 12 columns]

Анализ этих строк дает понять, что это данные о клиентах с пропущенной информацией о стаже и доходе. По сути, именно эти данные вместе могут обеспечивать уникальность записи по каждому клиенту,тогда как остальные (количество детей, пол, возраст, образование, тип занятости, семейное положение, цель кредита) вполне могут совпадать у разных клиентов, хоть и в небольшом количестве случаев. Поэтому можно сделать вывод, что эти дубликаты не случайны, а сформировались именно в результате остутствия информации по данным о стаже и доходе у ряда клиентов. В любом случае, для цели нашего исследования эти дубликаты значимости не имеют, поэтому мы можем спокойно их удалить с применением метода *drop_duplicates()* с созданием новых индексов:   

In [32]:
data = data.drop_duplicates().reset_index(drop = True)
data.duplicated().sum()

0

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

![](https://i.gifer.com/R0YH.gif)

### Лемматизация

Снова посмотрим значения столбца *'purpose'*, сформируем список его уникальных значений :

In [33]:
data_purpose = data['purpose'].unique()
data_purpose

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим список, который будет содержать леммы всех значений столбца *'purpose'*. Будем использовать библиотеку PyMystem:

In [34]:
from pymystem3 import Mystem
m = Mystem()

lemmas_purpose_list = []
for purpose in data_purpose:
    lemmas_purpose_list = lemmas_purpose_list + m.lemmatize(purpose)
#lemmas_purpose_list    
    
    

Используем коллекцию *Counter* для подсчета частоты встречаемости лемм в списке:

In [35]:
from collections import Counter
Counter(lemmas_purpose_list)

Counter({'покупка': 10,
         ' ': 59,
         'жилье': 7,
         '\n': 38,
         'приобретение': 1,
         'автомобиль': 9,
         'дополнительный': 2,
         'образование': 9,
         'сыграть': 1,
         'свадьба': 3,
         'операция': 4,
         'с': 5,
         'на': 4,
         'проведение': 1,
         'для': 2,
         'семья': 1,
         'недвижимость': 10,
         'коммерческий': 2,
         'жилой': 2,
         'строительство': 3,
         'собственный': 1,
         'подержать': 1,
         'свой': 4,
         'со': 1,
         'заниматься': 2,
         'сделка': 2,
         'подержанный': 1,
         'получение': 3,
         'высокий': 3,
         'профильный': 1,
         'сдача': 1,
         'ремонт': 1})

<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Реализация правильная, при этом предлагаю более короткое и быстрое решение с использованием [join()](https://stackoverflow.com/questions/5618878/how-to-convert-list-to-string)
```python
lemmas = m.lemmatize(' '.join(data['purpose']))
print(Counter(lemmas))
```
</div>

Проанализировав этот результат, мы можем выделить 5 основных лемм: *'недвижимость', 'автомобиль', 'свадьба', 'образование', 'ремонт'*. Все леммы типа *'жилье'* и *'жилой'* отнесем к категории *'недвижимость'*. Если бы в столбце *'purpose'* не было таких данных как *'недвижимость'* и *'образование'*, то мы бы могли расширить список категорий, указав еще отдельно типы недвижимости и вид образования, но такие неоднозначные данные присутствуют, значит сделать мы этого не можем. Таким образом, будем использовать словарь категорий: ***{'недвижимость', 'автомобиль', 'свадьба', 'образование', 'ремонт'}***.
Создадим функцию, которая будет определять категорию для каждого значения из столбца *'purpose'*. Для лемматизации удобно использовать библиотеку *PyMystem*:

In [36]:
def purpose_group(purpose):
    lemma_purpose = m.lemmatize(purpose)
    if ('недвижимость' in lemma_purpose) or ('жилье' in lemma_purpose) and ('ремонт' not in lemma_purpose):
        return 'недвижимость'
    elif 'автомобиль' in lemma_purpose:
        return 'автомобиль'
    elif 'свадьба' in lemma_purpose:
        return 'свадьба'
    elif 'образование' in lemma_purpose:
        return 'образование'
    elif 'ремонт' in lemma_purpose:
        return 'ремонт'

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Жирный плюс! Это один из сложнейших этапов проекта, справилась на отлично


Проверим, как работает функция, перебрав значения списка *data_purpose*:

In [37]:
for purpose in data_purpose:
    print(purpose_group(purpose))

недвижимость
автомобиль
образование
свадьба
недвижимость
образование
свадьба
недвижимость
недвижимость
недвижимость
недвижимость
недвижимость
недвижимость
недвижимость
автомобиль
автомобиль
недвижимость
недвижимость
недвижимость
недвижимость
автомобиль
образование
автомобиль
образование
автомобиль
свадьба
образование
недвижимость
недвижимость
образование
автомобиль
автомобиль
образование
образование
недвижимость
автомобиль
ремонт
образование


Наша функция работает корректно, можно применить ее к данным столбца *'purpose'* с помощью метода apply(). 

### Категоризация данных

In [38]:
data['purpose_group'] = data['purpose'].apply(purpose_group)
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_group  
0               покупка жилья  недвижимость  
1     приобретение автомобиля    автомобиль  
2               покупка жилья  недвижимость  
3  дополнительное образование   образование  
4             сыграть свадьбу       свадьба  
5               покупка жилья  недвижимость  
6           операции с жильем  недвижимость  
7                 образование   образование  
8       на проведение свадьбы       свадьба  
9     покупка жилья для семьи  недвижимость

Таким образом, мы создали дополнительный столбец *'purpose_group'*, в котором для каждого кредита указана нужная категория цели кредитования из словаря ***{'недвижимость', 'автомобиль', 'свадьба', 'образование', 'ремонт'}***. Это пригодится для дальнейшего проведения исследования.  

<div style="border:solid purple 5px; padding: 20px"> 
<h2 align="center"> Рубрика «Питонячий лайфхакер» <a class="tocSkip"> </h2>

<h3> Функция zip <a class="tocSkip"> </h3>

Функция zip создаёт итератор, который комбинирует элементы нескольких списков. Это позволяет осуществлять параллельный обход списков в циклах for или, например, выполнять параллельную сортировку.

![](https://i.ibb.co/MPPZ6TL/image.png)

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

Сгруппируем данные нашей таблицы по столбцу *'children'* и суммируем количество непогашенных кредитов по каждой группе:

In [39]:
data.groupby('children')['debt'].sum()

children
0     1063
1      445
2      194
3       27
4        4
5        0
20       8
Name: debt, dtype: int64

Используем эти данные для подсчета процентного отношения количества клиентов с задолженностью по возврату кредита к общему количеству клиентов по каждой группе. Для этого сначала посчитаем общее количество клиентов по каждой из групп,а потом найдем отношение :  

In [40]:
count = data.loc[data['children'] == 0,'children'].count()
print('Всего клиентов в категории без детей:',count)

Всего клиентов в категории без детей: 14091


In [41]:
procent = 1063/count
'Процент должников в категории без детей: {:.1%}'.format(procent)

'Процент должников в категории без детей: 7.5%'

In [42]:
count = data.loc[data['children'] == 1,'children'].count()
print('Всего клиентов в категории с 1 ребенком:',count)

Всего клиентов в категории с 1 ребенком: 4855


In [43]:
procent = 445/count
'Процент должников в категории с 1 ребенком: {:.1%}'.format(procent)

'Процент должников в категории с 1 ребенком: 9.2%'

In [44]:
count = data.loc[data['children'] == 2,'children'].count()
print('Всего клиентов в категории с 2 детьми:',count)

Всего клиентов в категории с 2 детьми: 2052


In [45]:
procent = 194/count
'Процент должников в категории с 2 детьми {:.1%}'.format(procent)

'Процент должников в категории с 2 детьми 9.5%'

In [46]:
count = data.loc[data['children'] == 3,'children'].count()
print('Всего клиентов в категории с 3 детьми:',count)

Всего клиентов в категории с 3 детьми: 330


In [47]:
procent = 27/count
'Процент должников в категории с 3 детьми {:.1%}'.format(procent)

'Процент должников в категории с 3 детьми 8.2%'

In [48]:
count = data.loc[data['children'] == 4,'children'].count()
print('Всего клиентов в категории с 4 детьми:',count)

Всего клиентов в категории с 4 детьми: 41


In [49]:
procent = 4/count
'Процент должников в категории с 4 детьми {:.1%}'.format(procent)

'Процент должников в категории с 4 детьми 9.8%'

In [50]:
count = data.loc[data['children'] == 5,'children'].count()
print('Всего клиентов в категории с 5 детьми:',count)

Всего клиентов в категории с 5 детьми: 9


In [51]:
procent = 0/count
'Процент должников в категории с 5 детьми {:.1%}'.format(procent)

'Процент должников в категории с 5 детьми 0.0%'

In [52]:
count = data.loc[data['children'] == 20,'children'].count()
print('Всего клиентов в категории с 20 детьми:',count)

Всего клиентов в категории с 20 детьми: 76


In [53]:
procent = 8/count
'Процент должников в категории с 20 детьми {:.1%}'.format(procent)

'Процент должников в категории с 20 детьми 10.5%'

<div class="alert alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Решение правильное, но я все же отмечу красным, чтобы ты могла оптимизировать решение. Такая реализация будет сложной, если в данных, по которым ты проводишь категоризацию, уникальных значений не 7, а, например, 50. Придется очень долго прописывать каждое условие. В идеале получить решение в сводной таблице, ты также для каждой группы можешь посчитать `count`, и там же поделить одно на другое (кстати, если ты в качестве агрегации возьмешь `mean` - получишь аналогичный результат). А если оставлять твое решение, то оберни его в цикл

<span style="color:magenta">
Предлагаю такой вариант:   
Для начала создадим функцию, которая будет принимать на вход нашу таблицу и название категорий, по которым ведется анализ (название столбца) и возвращать таблицу, содержащую следующую информацию:
столбец 'quantity_total' с общим количеством клиентов в категории;
столбец 'quantity_of_debts' с количеством должников по каждой категории;
столбец 'percent_of_debts' с долей должников в каждой категории в процентах:</span>

In [54]:
def ratio_result(data, categories):
    data_result = pd.DataFrame()
    data_result['quantity_total'] = data.groupby(categories)[categories].count()
    data_result['quantity_of_debts'] = data.groupby(categories)['debt'].sum()
    data_result['percent_of_debts'] = round(data_result['quantity_of_debts']/data_result['quantity_total']*100,1)
    return data_result

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 <a class="tocSkip"> </h2>

Классное решение, а можно было вот так

In [89]:
data.groupby('children')['debt'].agg(['count', 'sum', 'mean']).style.format({'mean': '{:.2%}'})

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 <a class="tocSkip"> </h2>

Или так

In [91]:
data.pivot_table(index='children', values='debt', aggfunc=['count', 'sum', 'mean']).style.format({('mean', 'debt'): '{:.2%}'})

<span style="color:magenta">Применим эту функцию к столбцу 'children':</span>

In [55]:
ratio_result(data, 'children')

quantity_total  quantity_of_debts  percent_of_debts
children                                                     
0                  14091               1063               7.5
1                   4855                445               9.2
2                   2052                194               9.5
3                    330                 27               8.2
4                     41                  4               9.8
5                      9                  0               0.0
20                    76                  8              10.5

**Вывод:**
***
Всего клиентов без детей: 14091      
Процент должников в категории без детей: 7.5%
***
Всего клиентов с 1 ребенком: 4855   
Процент должников в категории с 1 ребенком: 9.2%
***
Всего клиентов с 2 детьми: 2052      
Процент должников в категории с 2 детьми 9.5%
***
Всего клиентов с 3 детьми: 330       
Процент должников в категории с 3 детьми 8.2%
***
Всего клиентов с 4 детьми: 41        
Процент должников в категории с 4 детьми 9.8%
***
Всего клиентов с 5 детьми: 9         
Процент должников в категории с 5 детьми 0.0%
***
Всего клиентов с 20 детьми: 76  
Процент должников в категории с 20 детьми 10.5%

Просмотр результатов дает понять, что с ростом количества детей в семье возрастает относительное количество непогашенных кредитов. Исключение составляет случай с 3 детьми в семье, когда процент должников  немного падает по сравнению с группой с 2 детьми, и с 5 детьми в семье, где непогашенных кредитов нет вообще, что может быть связано с недостаточным обьемом выборки по этой группе клиентов (всего 9 семей с 5 детьми). Самый большой процент должников в группе семей с 20 детьми (более 10% таких клиентов не погашают кредиты в срок). Несмотря на то что большая часть непогашенных кредитов (1063) приходится на клиентов без детей, за счет их подавляющего количества, только около 7.5% таких клиентов оказывается в должниках, что позволяет считать их самыми надежными клиентами банка, наряду с клиентами у которых 3 детей (процент должников 8.2%).

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

Сгруппируем данные нашей таблицы по столбцу *'family_status'* и суммируем количество непогашенных кредитов по каждой группе:

In [56]:
data.groupby('family_status')['debt'].sum()

family_status
в разводе                 85
вдовец / вдова            63
гражданский брак         388
женат / замужем          931
не женат / не замужем    274
Name: debt, dtype: int64

Используем эти данные для подсчета процентного отношения количества клиентов с задолженностью по возврату кредита к общему количеству клиентов по каждой группе. Для этого сначала посчитаем общее количество клиентов по каждой из групп,а потом найдем отношение :

In [57]:
count = data.loc[data['family_status'] == 'в разводе','family_status'].count()
print('Всего клиентов со статусом "в разводе":',count)

Всего клиентов со статусом "в разводе": 1195


In [58]:
procent = 85/count
'Процент должников в категории со статусом "в разводе": {:.1%}'.format(procent)

'Процент должников в категории со статусом "в разводе": 7.1%'

In [59]:
count = data.loc[data['family_status'] == 'вдовец / вдова','family_status'].count()
print('Всего клиентов со статусом "вдовец / вдова":',count)

Всего клиентов со статусом "вдовец / вдова": 959


In [60]:
procent = 63/count
'Процент должников в категории со статусом "вдовец / вдова": {:.1%}'.format(procent)

'Процент должников в категории со статусом "вдовец / вдова": 6.6%'

In [61]:
count = data.loc[data['family_status'] == 'гражданский брак','family_status'].count()
print('Всего клиентов со статусом "гражданский брак":',count)

Всего клиентов со статусом "гражданский брак": 4151


In [62]:
procent = 388/count
'Процент должников в категории со статусом "гражданский брак": {:.1%}'.format(procent)

'Процент должников в категории со статусом "гражданский брак": 9.3%'

In [63]:
count = data.loc[data['family_status'] == 'женат / замужем','family_status'].count()
print('Всего клиентов со статусом "женат / замужем":',count)

Всего клиентов со статусом "женат / замужем": 12339


In [64]:
procent = 931/count
'Процент должников в категории со статусом "женат / замужем": {:.1%}'.format(procent)

'Процент должников в категории со статусом "женат / замужем": 7.5%'

In [65]:
count = data.loc[data['family_status'] == 'не женат / не замужем','family_status'].count()
print('Всего клиентов со статусом "не женат / не замужем":',count)

Всего клиентов со статусом "не женат / не замужем": 2810


In [66]:
procent = 274/count
'Процент должников в категории со статусом "не женат / не замужем": {:.1%}'.format(procent)

'Процент должников в категории со статусом "не женат / не замужем": 9.8%'

<span style="color:magenta">Задача похожа на предыдущую. Оптимизируем решение - снова применим функцию 'ratio_result', на этот раз к столбцу 'family_status':</span>

In [67]:
ratio_result(data,'family_status')

quantity_total  quantity_of_debts  percent_of_debts
family_status                                                             
в разводе                        1195                 85               7.1
вдовец / вдова                    959                 63               6.6
гражданский брак                 4151                388               9.3
женат / замужем                 12339                931               7.5
не женат / не замужем            2810                274               9.8

**Вывод:**
***
Всего клиентов со статусом "в разводе": 1195   
Процент должников в категории со статусом "в разводе": 7.1%
***
Всего клиентов со статусом "вдовец / вдова": 959  
Процент должников в категории со статусом "вдовец / вдова": 6.6%
***
Всего клиентов со статусом "гражданский брак": 4151  
Процент должников в категории со статусом "гражданский брак": 9.3%
***
Всего клиентов со статусом "женат / замужем": 12339  
Процент должников в категории со статусом "женат / замужем": 7.5%
***
Всего клиентов со статусом "не женат / не замужем": 2810  
Процент должников в категории со статусом "не женат / не замужем": 9.8%

Самый большой процент должников в группе "не женат / не замужем" - 9.8% таких клиентов не гасят кредиты в срок, это самые ненадежные клиенты. Самые надежные клиенты с точки зрения своевременного погашения кредита - клиенты с семейным статусом "вдовец / вдова"  (только 6.6% должников в своей категории) и со статусом "в разводе" (7.1% должников в своей категории).  

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Сгруппируем данные нашей таблицы по столбцу  *'total_income'* и суммируем количество непогашенных кредитов по каждой группе. Результат сохраним в новой таблице data_group:

In [68]:
data_group = data.groupby('total_income')['debt'].sum()
data_group.head(10)

total_income
20667    1
21205    0
21367    0
21695    0
21895    0
22472    0
23844    0
24457    0
25227    0
25308    0
Name: debt, dtype: int64

Отсортируем данные новой таблицы по убыванию:

In [69]:
data_group.sort_values(ascending = False).head(10)

total_income
142594    98
118514    35
172357    30
150447     7
83436      2
132113     2
113004     2
20667      1
97788      1
97793      1
Name: debt, dtype: int64

Определим сколько всего непогашенных кредитов у всех клиентов:

In [70]:
data['debt'].sum()

1741

Посчитаем процентное отношение должников - 'лидеров' по категории 'размер дохода', данные возьмем из предыдущего результата с сортировкой: 

In [71]:
'{:.1%}'.format((98 + 35 + 30 + 7)/1741)

'9.8%'

Как видим, 9.8 % непогашенных кредитов приходится на клиентов с медианными значением дохода, которые мы определяли выше. По сути значительная часть должников приходится на клиентов, у которых были пропущены данные о доходе и общем стаже. Можно предположить, что эти пропуски не случайны и связаны с действительным отсутствием постоянного источника дохода у этих клиентов. Остальные случаи непогашения единичны и не связаны напрямую с величиной дохода.

<div class="alert alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

Интересное решение 🤔 А попробуй предварительно категоризовать данные по доходу. Определи несколько отсечек для деления (например по процентилям), определи новый столбец с категорией по доходу и посмотри, как меняется уровень должников по этим группам. Здесь, кстати для деления можно применить метод [cut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html) или [qcut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html)(в этом случае деление будет проводиться по проценилям). Обрати внимание, у данного метода есть параметр `labels`, так можно оформить группы красивее 😊  

<span style="color:magenta">Хороший совет, постараюсь реализовать. Данные по доходу можно категоризовать следующим образом. Разделим весь диапазон доходов на 4 равные части (квартили). Для наглядности создадим список названий этих частей - это и будет словарь категорий:</span>

In [72]:
labels_list = ['quartile1','quartile2','quartile3','quartile4']

<span style="color:magenta">С помощью метода qcut создадим столбец 'total_income_quartile', в котором и будут определены категории данных по доходу. Используем также метод describe() для наглядного представления наших диапазонов-квартилей:</span> 

In [73]:
data['total_income_quartile'] = pd.qcut(data['total_income'], q = 4, labels = labels_list)
data['total_income'].describe()

count    2.145400e+04
mean     1.653196e+05
std      9.818730e+04
min      2.066700e+04
25%      1.076230e+05
50%      1.425940e+05
75%      1.958202e+05
max      2.265604e+06
Name: total_income, dtype: float64

<span style="color:magenta">Теперь мы снова можем обратиться к нашей функции ratio_result, применив ее к столбцу 'total_income_quartile':</span> 

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 <a class="tocSkip"> </h2>

Отличная реализация 👏

In [74]:
ratio_result(data, 'total_income_quartile')

quantity_total  quantity_of_debts  percent_of_debts
total_income_quartile                                                     
quartile1                        5364                427               8.0
quartile2                        5479                483               8.8
quartile3                        5247                448               8.5
quartile4                        5364                383               7.1

<span style="color:magenta">**Вывод**  
Как можно заметить, распределение доли должников по категориям 'размер дохода' почти равномерное. Самый большой процент непогашенных кредитов - 8.8% приходится на клиентов с уровнем дохода,находящимся во втором квартиле (диапазон дохода (107623,142594]). Самый маленький процент непогашенных кредитов - 7.1% - находится в четвертом квартиле доходов (диапазон (195820,2265604]). В основном надежность клиента не зависит от величины его дохода, скорее всего на это больше влияют другие факторы.</span>   

**Вывод**

В около 10 % случаев основными должниками являются клиенты, которые не указали свой доход и трудовой стаж, т.е.предположительно безработные или официально не трудоустроеные, не имеющие постоянного дохода. Остальные случаи непогашения в срок равномерно распределены по остальным клиентам и не связаны напрямую с величиной дохода.

- Как разные цели кредита влияют на его возврат в срок?

Сгруппируем данные нашей таблицы по столбцу *'purpose_group'* и суммируем количество непогашенных кредитов по каждой группе:

In [75]:
data.groupby('purpose_group')['debt'].sum()

purpose_group
автомобиль      403
недвижимость    747
образование     370
ремонт           35
свадьба         186
Name: debt, dtype: int64

Используем эти данные для подсчета процентного отношения количества просроченных кредитов к общему количеству кредитов по каждой группе. Для этого сначала посчитаем общее количество кредитов по каждой из групп,а потом найдем отношение :

In [76]:
count = data.loc[data['purpose_group'] == 'автомобиль','purpose_group'].count()
print('Всего кредитов в категории "автомобиль": ',count)

Всего кредитов в категории "автомобиль":  4306


In [77]:
procent = 403/count
'Процент просроченных кредитов в категории "автомобиль" : {:.1%}'.format(procent)

'Процент просроченных кредитов в категории "автомобиль" : 9.4%'

In [78]:
count = data.loc[data['purpose_group'] == 'недвижимость','purpose_group'].count()
print('Всего кредитов в категории "недвижимость": ',count)

Всего кредитов в категории "недвижимость":  10204


In [79]:
procent = 747/count
'Процент просроченных кредитов в категории "недвижимость" : {:.1%}'.format(procent)

'Процент просроченных кредитов в категории "недвижимость" : 7.3%'

In [80]:
count = data.loc[data['purpose_group'] == 'образование','purpose_group'].count()
print('Всего кредитов в категории "образование": ',count)

Всего кредитов в категории "образование":  4013


In [81]:
procent = 370/count
'Процент просроченных кредитов в категории "образование" : {:.1%}'.format(procent)

'Процент просроченных кредитов в категории "образование" : 9.2%'

In [82]:
count = data.loc[data['purpose_group'] == 'ремонт','purpose_group'].count()
print('Всего кредитов в категории "ремонт": ',count)

Всего кредитов в категории "ремонт":  607


In [83]:
procent = 35/count
'Процент просроченных кредитов в категории "ремонт" : {:.1%}'.format(procent)

'Процент просроченных кредитов в категории "ремонт" : 5.8%'

In [84]:
count = data.loc[data['purpose_group'] == 'свадьба','purpose_group'].count()
print('Всего кредитов в категории "свадьба": ',count)

Всего кредитов в категории "свадьба":  2324


In [85]:
procent = 186/count
'Процент просроченных кредитов в категории "свадьба" : {:.1%}'.format(procent)

'Процент просроченных кредитов в категории "свадьба" : 8.0%'

<span style="color:magenta">Оптимизируем решение - снова применим функцию 'ratio_result', на этот раз к столбцу 'purpose_group':</span>

In [86]:
ratio_result (data, 'purpose_group')

quantity_total  quantity_of_debts  percent_of_debts
purpose_group                                                     
автомобиль               4306                403               9.4
недвижимость            10204                747               7.3
образование              4013                370               9.2
ремонт                    607                 35               5.8
свадьба                  2324                186               8.0

**Вывод:**
***
Процент просроченных кредитов в категории "автомобиль" : 9.4%
***
Процент просроченных кредитов в категории "образование" : 9.2%
***
Процент просроченных кредитов в категории "свадьба" : 8.0%
***
Процент просроченных кредитов в категории "недвижимость" : 7.3%
***
Процент просроченных кредитов в категории "ремонт" : 5.8%

Самая рискованная цель кредитования с точки зрения возврата в срок - на покупку автомобиля, 9.4% таких кредитов не гасят своевременно. Самая безопасная - на ремонт, там доля просроченных кредитов составляет только 5.8%. 

<div style="border:solid purple 5px; padding: 20px"> 
<h2 align="center"> Рубрика «Питонячий лайфхакер» <a class="tocSkip"> </h2>

<h3> Встроенные функции all() и any() <a class="tocSkip"> </h3>

Полезные функции, которые помогут узнать есть ли определенные значения в объекте.

`all` возвращает **True**, если все объекты равны **True** или удовлетворяют условию

![](https://i.ibb.co/NKs2yWq/image.png)

`any` в свою очередь делает то же самое, но возвращает **True**,если хотя бы один равен **True**

![](https://i.ibb.co/4tq8TmV/image.png)

## Шаг 4. Общий вывод

В качестве входных данных была получена таблица с информацией о платежеспособности клиентов. Данные были изучены на предмет определения пропущенных значений и дубликатов. Пропущенные значения были заполнены таким образом, чтобы не нарушать целостность выборки при дальнейшем анализе данных (использовались медианные значения для отдельных категорий клиентов) . Дубликаты были изучены и удалены, их удаление не могло повлиять на статистику дальнейших выводов. Была проведена категоризация данных по целям кредитования,что позволило провести эффективный анализ зависимости платежеспособности клиентов от этих целей. Для наглядности выбора категорий производилась лемматизация данных. 
В качестве ответов на поставленные вопросы были произведены статистические расчеты и результаты представлены в виде процентного отношения доли непогашенных кредитов в той или иной категории клиентов или целей кредитования.      

<div style="border:solid purple 2px; padding: 20px"> 

У тебя получилась очень хорошая работа! Многие моменты сделаны просто прекрасно. Ты показываешь отличное владение изучаемым материалом: уверенно пользуешься pandas, умеешь подготавливать, очищать, обогощать данные. На протяжении всей работы чувствуется глубина проработки задачи и это безусловный плюс для нас.

Остались небольшие доработки:

- В шаге 2 необходимо провести категоризацию данных по уровню дохода;
- В шаге 3 необходимо оптимизировать решение;

Жду твоих исправлений, у тебя все получится 😊

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 <a class="tocSkip"> </h2>

Спасибо за быструю и качественную доработку комментариев! Все моменты были исправлены, так что работу я принимаю. Надеюсь ты учтешь мои комментарии в последующих работах. Благодарю за старания и интерес и желаю успехов в дальнейших проектах 😊

</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.